### DISCUSSION TITLE
Code structure for IoT device to write to Supabase

### DISCUSSION THREAD

**Discussion by meierthomas** [supabase/supabase#3805] [General]
I'm more of a maker rather than a developer so please bear with me if this is super basic :)

I'm putting together a prototype for a battery powered IoT sensor that regularly writes sensor data to Supabase. On the other hand I have a web app to access this data in Supabase for analysis etc. (used by the user who installed the IoT sensor). What's not clear in the context of RLS is how I insert the data against the specific user.

Using the Supabase APIs I'm able to successfully post sensor values to the DB (RLS disabled).

Looking at the documentation, Supabase assigns a unique `user_id` when a user signs up.

This is how I think the flow could be like:
1. User creates a login on the web app
2. User sets up the IoT sensor; Wifi credentials and 'Supabase ID'
3. IoT device posts data via the Supabase APIs and 'Supabase ID'

What's not clear is what 'Supabase ID' does the user provide in step 2. Looking at the documentation the user should provide `user_id` which is not known (all they have is the login - e.g. email address). Is it possible to lookup `user_id` by providing the web app login? If yes `user_id` can be stored on the IoT sensor and used for database writes

**Comment by zakaria-chahboun** [comment]
First thing, Your user has to authenticate.

Suppose you have this *public.users* table instruction:

|user_id|battery_status|..|
|------|------|-----|
|9e2a3d7a|20%|..|
|b7607d68|95%|..|
|c657751a|44%|..|

You just create a `RLS` to avoid overwriting all users data:
```SQL
CREATE POLICY allow_user ON public.users
USING (auth.uid() = user_id);
```

The `auth.uid()` is *the current authenticated user id*. PostgreSQL know him from the `JSON token` automatically!

> Now. only the current user can *INSERT DELETE SELECT UPDATE* his row in the *public.users*

**Comment by GaryAustin1** [comment]
You might consider either using an ID (could be a UUID) from the device, or a UUID code a logged in user can get for the device and then go set it in device.  So user has his normal login ID, but each device has it's own ID also.  There is a table that has the user ID and device ID in it.  Now each device is linked to a user and each user can have one or more devices in this link table. 
 
You then use RLS or RPC function to check when the device writes to it's table if the device ID is in the link table to stop bogus writes.  The user can also read the device table by providing device ID and use RLS or RPC to check the link table that the user has access to that device.

With this your devices don't actually have to login and deal with tokens etc.  All they can do is write the device info to a row/s with their device number and they must have already been registered in the join table by the user.

**Reply by meierthomas** [reply]
I would use a device UUID (e.g. Mac address etc.). Are you suggesting to register the device it connects to Supabase and writes the device UUID and the login ID (let's assume the login ID is stored on the device) to the link table? To do this I use the apikey, authorisation and the rest of the headers to create a http POST together with the relevant payload (being the device UUID and the login ID) to write to the link table?

Then there's a second table that holds the sensor values, the device UUID it came from and the login ID and the app will be able to look up all devices registered to the login ID via the link table

**Reply by GaryAustin1** [reply]
So the login ID in my concept is the user (persons) id.  The person sets up an account and sets the id of the device (or devices) he has access to.  That relationship user id/ device id is stored in a table with user ID RLS.   Then there is a table that is public on for writing, but you must have a device id to set with the row, and then the user id (with an additional check in the link table for user/device ids both being in a row) for reading that has that sensor data.  The user can only see his own sensors.  The sensor never logs in, and just does a write to the sensor data table with it's unique id (mac).  The user has to log in to read the sensor data or manage devices, but can only read ones he set up.

**Reply by GaryAustin1** [reply]
add .... the write to sensor data table should probably RLS check the device id exists in the link table also to stop wasted writes by a bad actor.  Then you would have to allow RLS access to read link table on user ID for uid column or device id in did column.  For even more security you could add a device password column that the user sets in both the device and in the app if you are worried about someone guessing a mac address and writing fake data.  To me an assigned UID that gets set to the device instead of mac would be as good as a password.
add again...  Basically I'm just proposing you not login devices, but people with devices.  Then pass what ever security info from  the device in the write each time, versus logging devices in and not having to deal with that process, tokens,auth ids, etc. in the device.

**Reply by meierthomas** [reply]
Ok, I'm a visual person. So this is what I picked up from your response and I hope this makes sense. One thing that seems to be clear is the sensor module only writes to table2 with device_id and login_id while the app writes to table1 and reads from table2.

An additional challenge is the sensor module is battery powered and in deepsleep for most of the time. It wakes up at regular intervals and writes data to table2. As such the MAC is still the safest option as the user picks up the MAC from a sticker on the sensor module, enters it in the app together with the login_id (email address) while the login_id on the sensor module can be entered when the sensor module first gets setup as part of the connection to Wifi. The device_id can be generated automatically from the sensor module.

Not sure where I would put your statement "Then you would have to allow RLS access to read link table on user ID for uid column or device id in did column"

![image](https://user-images.githubusercontent.com/24826786/141984829-8b385245-c948-49d2-96e5-e6a1093f6454.png)

Am I on the right track here?

**Reply by GaryAustin1** [reply]
Yes, but security needs to be sorted out.  First login_id I would use uid just for simplicity but you could use email.
I don't know your security needs for the device writes.  Technically if you are not worried about someone faking one of the devices, you don't need the login id on the device.  You just make sure you have a write RLS set on table2 to check the MAC is in table 1.  You also need a read RLS on table 2 to check if login_id has a device_id match in table 1 and then read table 2 to just by device_id.    No login ID needed in table 2.  Table 1 RLS is standard uuid for reads and writes so only owner has access. If you do the device write security check then you also need the read RLS on table 1 to allow for device_id match.

If you need more security then you can use any random number appended to the mac as the device_id.  Just like a password (but no need to encrypt).   So device_id is something like MAC-ABCDEF and you have pretty tough device_id to guess.  The user can program the extra info into the device.  Your RLS would be the same as above for table 2.   

Note you can expand this so 1 user can have multiple devices by allowing extra entries in table 1 (and of course having interface to deal with it in app.  In this case primary key is auto generated and then the device_id and login_id columns.  If you need lots of entries in table 2 for each device then you should also just have an auto generated primary key and then device_id,valuea, valueb.

I made a couple of fixes above in edit.

**Reply by GaryAustin1** [reply]
Sorry my responses get long....  In thinking about table 1 security I would have RLS only allowing owner read.  Then I would have a Postgres function set as security definer (by pass RLS) to check if device_id is in table 1 for use in the device write RLS on table 2.

**Reply by GaryAustin1** [reply]
This is an example of a function I use for a similar type RLS check arrangement (chat room with user relationship table) to test in RLS on the room table.

create or replace function public.check_room_member(roomkey varchar, userid uuid) 
returns boolean as $$
begin
SELECT exists(
  SELECT 1 FROM room_joins WHERE room_key=roomkey AND uid=userid
  );
end;
$$ language plpgsql security definer stable;

**Comment by zmzlois** [comment]
> I'm more of a maker rather than a developer so please bear with me if this is super basic :)
> 
> I'm putting together a prototype for a battery powered IoT sensor that regularly writes sensor data to Supabase. On the other hand I have a web app to access this data in Supabase for analysis etc. (used by the user who installed the IoT sensor). What's not clear in the context of RLS is how I insert the data against the specific user.
> 
> Using the Supabase APIs I'm able to successfully post sensor values to the DB (RLS disabled).
> 
> Looking at the documentation, Supabase assigns a unique `user_id` when a user signs up.
> 
> This is how I think the flow could be like:
> 
> 1. User creates a login on the web app
> 2. User sets up the IoT sensor; Wifi credentials and 'Supabase ID'
> 3. IoT device posts data via the Supabase APIs and 'Supabase ID'
> 
> What's not clear is what 'Supabase ID' does the user provide in step 2. Looking at the documentation the user should provide `user_id` which is not known (all they have is the login - e.g. email address). Is it possible to lookup `user_id` by providing the web app login? If yes `user_id` can be stored on the IoT sensor and used for database writes

@meierthomas can I ask how did you write sensor data to supabase DB? is their any instruction I can follow?

**Comment by meierthomas** [comment]
Sorry, I have put this project on hold for a few reasons so unfortunately I
can't help you at this time. Will come back to this but don't know exactly
when. Hopefully you find your answer soon.
